File: packages/media_processing_kit/android/src/main/kotlin/com/bittercamera/media_processing_kit/PassthroughEffectProcessor.kt
package com.bittercamera.media_processing_kit

import java.io.File
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean

/**
 * 默认内置的 effect：passthrough
 *
 * 人话解释：
 * - 先不做任何“苦瓜脸”处理
 * - 只把文件复制一份，并按进度回调 0 -> 1
 * - 用于让 A/C/E 先跑通 D 黑盒的任务管理/进度/取消/错误码
 */
class PassthroughEffectProcessor : EffectProcessor {
    private val cancelled = ConcurrentHashMap<String, AtomicBoolean>()

    override fun processVideo(
        taskId: String,
        inputPath: String,
        outputPath: String,
        config: Map<String, Any?>,
        callbacks: TaskCallbacks
    ) {
        cancelled[taskId] = AtomicBoolean(false)
        try {
            callbacks.onProgress(taskId, 0.0)
            if (cancelled[taskId]?.get() == true) {
                callbacks.onError(taskId, "VIDEO_PROCESS_CANCELLED", "Cancelled")
                return
            }
            val inFile = File(inputPath)
            val outFile = File(outputPath)
            outFile.parentFile?.mkdirs()
            inFile.copyTo(outFile, overwrite = true)
            callbacks.onProgress(taskId, 1.0)
            callbacks.onCompleted(taskId, outFile.absolutePath)
        } catch (e: Exception) {
            callbacks.onError(taskId, "UNKNOWN", e.message ?: "unknown")
        } finally {
            cancelled.remove(taskId)
        }
    }

    override fun processImage(
        taskId: String,
        inputPath: String,
        outputPath: String,
        config: Map<String, Any?>,
        callbacks: TaskCallbacks
    ) {
        // 图片也先做 copy，便于联调
        processVideo(taskId, inputPath, outputPath, config, callbacks)
    }

    override fun cancel(taskId: String) {
        cancelled[taskId]?.set(true)
    }
}

File: packages/media_processing_kit/ios/Classes/PassthroughEffectProcessor.swift
import Foundation

/// 默认内置 passthrough effect：
/// - 仅 copy 文件，用于联调任务/进度/取消/事件回调
public final class PassthroughEffectProcessor: EffectProcessor {
  private var cancelled: Set<String> = []
  private let lock = NSLock()

  public init() {}

  public func processVideo(taskId: String, inputPath: String, outputPath: String, config: [String : Any?], callbacks: TaskCallbacks) {
    callbacks.onProgress(taskId: taskId, progress: 0.0)
    if isCancelled(taskId: taskId) {
      callbacks.onError(taskId: taskId, errorCode: "VIDEO_PROCESS_CANCELLED", errorMessage: "Cancelled")
      return
    }

    do {
      let fm = FileManager.default
      let outURL = URL(fileURLWithPath: outputPath)
      try fm.createDirectory(at: outURL.deletingLastPathComponent(), withIntermediateDirectories: true)
      if fm.fileExists(atPath: outputPath) {
        try fm.removeItem(atPath: outputPath)
      }
      try fm.copyItem(atPath: inputPath, toPath: outputPath)
      callbacks.onProgress(taskId: taskId, progress: 1.0)
      callbacks.onCompleted(taskId: taskId, outputPath: outputPath)
    } catch {
      callbacks.onError(taskId: taskId, errorCode: "UNKNOWN", errorMessage: error.localizedDescription)
    }
    clear(taskId: taskId)
  }

  public func processImage(taskId: String, inputPath: String, outputPath: String, config: [String : Any?], callbacks: TaskCallbacks) {
    processVideo(taskId: taskId, inputPath: inputPath, outputPath: outputPath, config: config, callbacks: callbacks)
  }

  public func cancel(taskId: String) {
    lock.lock(); defer { lock.unlock() }
    cancelled.insert(taskId)
  }

  private func isCancelled(taskId: String) -> Bool {
    lock.lock(); defer { lock.unlock() }
    return cancelled.contains(taskId)
  }

  private func clear(taskId: String) {
    lock.lock(); defer { lock.unlock() }
    cancelled.remove(taskId)
  }
}

